memory_profiler.py

```diff
@@ -593,209 +603,271 @@ def show_results(prof, stream=None, precision=1):
-
-    mem_usage = 0
-    counter = 0
-    baseline = memory_usage()[0]
-    while counter < repeat:
-        counter += 1
-        tmp = memory_usage((_func_exec, (stmt, self.shell.user_ns)),
-                           timeout=timeout, interval=interval, max_usage=True,
-                           include_children=include_children)
-        mem_usage = max(mem_usage, tmp[0])
-
-    if mem_usage:
-        print('peak memory: %.02f MiB, increment: %.02f MiB' %
-              (mem_usage, mem_usage - baseline))
-    else:
-        print('ERROR: could not read memory usage, try with a lower interval '
-              'or more iterations')
+          In [3]: %%memit l=range(1000000)
+             ...: len(l)
+             ...:
+          peak memory: 52.14 MiB, increment: 0.08 MiB
+
+        """
+        from memory_profiler import memory_usage, _func_exec
+        opts, stmt = self.parse_options(line, 'r:t:i:c', posix=False, strict=False)
+
+        if cell is None:
+            setup = 'pass'
+        else:
+            setup = stmt
+            stmt = cell
+
+        repeat = int(getattr(opts, 'r', 1))
+        if repeat < 1:
+            repeat == 1
+        timeout = int(getattr(opts, 't', 0))
+        if timeout <= 0:
+            timeout = None
+        interval = float(getattr(opts, 'i', 0.1))
+        include_children = 'c' in opts
+
+        # I've noticed we get less noisier measurements if we run
+        # a garbage collection first
+        import gc
+        gc.collect()
+
+        _func_exec(setup, self.shell.user_ns)
+
+        mem_usage = 0
+        counter = 0
+        baseline = memory_usage()[0]
+        while counter < repeat:
+            counter += 1
+            tmp = memory_usage((_func_exec, (stmt, self.shell.user_ns)),
+                               timeout=timeout, interval=interval, max_usage=True,
+                               include_children=include_children)
+            mem_usage = max(mem_usage, tmp[0])
+
+        if mem_usage:
+            print('peak memory: %.02f MiB, increment: %.02f MiB' %
+                  (mem_usage, mem_usage - baseline))
+        else:
+            print('ERROR: could not read memory usage, try with a lower interval '
+                  'or more iterations')
+
+    @classmethod
+    def register_magics(cls, ip):
+        from distutils.version import LooseVersion
+        import IPython
+        ipython_version = LooseVersion(IPython.__version__)
+
+        if ipython_version < '0.13':
+            try:
+                _register_magic = ip.define_magic
+            except AttributeError:  # ipython 0.10
+                _register_magic = ip.expose_magic
+
+            _register_magic('mprun', cls.mprun.__func__)
+            _register_magic('memit', cls.memit.__func__)
+        else:
+            ip.register_magics(cls)
+
+# Ensuring old interface of magics expose for IPython 0.10
+magic_mprun = MemoryProfilerMagics.mprun.__func__
+magic_memit = MemoryProfilerMagics.memit.__func__
```

Comment:
@korzeniewskipl shouldn't this be ```MemoryProfilerMagics().mprun.__func__``` (note the parenthesis) or just ```MemoryProfilerMagics.mprun```? Otherwise I'm having exceptions under Python3 ...